crates/polars-time/src/windows/duration.rs

```diff
@@ -487,40 +487,52 @@ impl Duration {
         &self,
         t: i64,
         tz: Option<&Tz>,
-        timestamp_to_datetime: G,
-        datetime_to_timestamp: J,
+        _timestamp_to_datetime: G,
+        _datetime_to_timestamp: J,
+        daily_duration: i64,
     ) -> PolarsResult<i64>
     where
         G: Fn(i64) -> NaiveDateTime,
         J: Fn(NaiveDateTime) -> i64,
     {
-        let _original_dt_utc;
-        let original_dt_naive;
-        let dt = match tz {
+        let _original_dt_utc: Option<NaiveDateTime>;
+        let _original_dt_local: Option<NaiveDateTime>;
+        let t = match tz {
             #[cfg(feature = "timezones")]
             Some(tz) => {
-                _original_dt_utc = timestamp_to_datetime(t);
-                original_dt_naive = unlocalize_datetime(_original_dt_utc, tz);
-                original_dt_naive.date()
+                _original_dt_utc = Some(_timestamp_to_datetime(t));
+                _original_dt_local = Some(unlocalize_datetime(_original_dt_utc.unwrap(), tz));
+                _datetime_to_timestamp(_original_dt_local.unwrap())
             },
             _ => {
-                _original_dt_utc = timestamp_to_datetime(t);
-                original_dt_naive = _original_dt_utc;
-                original_dt_naive.date()
+                _original_dt_utc = None;
+                _original_dt_local = None;
+                t
             },
         };
-        let week_timestamp = dt.week(Weekday::Mon);
-        let first_day_of_week =
-            week_timestamp.first_day() - chrono::Duration::weeks(self.weeks - 1);
-        let result_dt_naive = first_day_of_week.and_time(NaiveTime::default());
+        // If we did
+        //   t - (t % (7 * self.weeks * daily_duration))
+        // then the timestamp would get truncated to the previous Thursday,
+        // because 1970-01-01 (timestamp 0) is a Thursday.
+        // So, we adjust by 4 days to get to Monday.
+        let mut remainder = (t - 4 * daily_duration) % (7 * self.weeks * daily_duration);
+        if remainder < 0 {
+            remainder += 7 * self.weeks * daily_duration
+        }
+        let result_t_local = t - remainder;
```

Comment:
This is the gist of the change

Instead of using Chrono to go to the previous monday, just do the arithmetic on the underlying integers - this matches the pattern used in `truncate_subweekly`